### ISSUE TITLE
Docker 1.10 : Using WORKDIR with non-existing path results in container terminating with SIGSEGV

### ISSUE THREAD

**Issue by nick-pap** [docker/docker#21082]
Output of `docker version`:

Client:
 Version:      1.10.0
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   590d5108
 Built:        Thu Feb  4 18:34:50 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.0
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   590d5108
 Built:        Thu Feb  4 18:34:50 2016
 OS/Arch:      linux/amd64

Output of `docker info`:

Containers: 13
 Running: 2
 Paused: 0
 Stopped: 11
Images: 428
Server Version: 1.10.0
Storage Driver: devicemapper
 Pool Name: docker-253:1-167772291-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: ext4
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 8.51 GB
 Data Space Total: 107.4 GB
 Data Space Available: 98.86 GB
 Metadata Space Used: 12.72 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.135 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Either use `--storage-opt dm.thinpooldev` or use `--storage-opt dm.no_warn_on_loop_devices=true` to suppress this warning.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.107-RHEL7 (2015-10-14)
Execution Driver: native-0.2
Logging Driver: json-file
Plugins:
 Volume: local
 Network: bridge null host
Kernel Version: 3.10.0-327.3.1.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 23.39 GiB
Name: bambdev
ID: BSEV:TY3M:YQSN:GXNQ:BLSO:GNTP:BG4S:UIBK:7OLI:UN4D:AKO2:77IL
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled

Provide additional environment details (AWS, VirtualBox, physical, etc.):

VirtualBox VM

List the steps to reproduce the issue:
1. Create a Dockerfile with the following contents:
--------- Dockerfile ---------------
`FROM ubuntu:15.04`
`ENV AP /app`
`WORKDIR $AP/`
`ADD dummyfile.txt $AP/`
`CMD ["ls"]`
--------- EOF Dockerfile ---------------
2. Build the docker image: `docker build -t test .`
3. Run the docker image: `docker run -it test` and note that there is no screen output while we would expect to see "dummyfile.txt" shown.
4. Do a `docker ps -a` and notice that the container exited with code 139 (SIGSEGV).
5. Do a `docker logs test` and notice that nothing has been output to the log file.
6. In the Dockerfile move the `WORKDIR` directive below the `ADD` directive and repeat from step 2. Notice that this time, the `ls` command is executed and the correct result is displayed "dummyfile.txt".

The script from step 1 worked as expected at least until:
Client:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:16:54 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:16:54 UTC 2015
 OS/Arch:      linux/amd64

**Comment by WeiZhang555** [comment]
Have you tested it on master branch? I can't reproduce it on master

**Comment by nick-pap** [comment]
Not tested on master branch. On the machine this happens I don't have access to build/install Docker.

**Comment by nick-pap** [comment]
Update:
TL/DR : The user had the bug...

I've been using several scripts and a Dockerized GCC 5.2 to build some binaries and then pack them into another Docker image. The actual script contained something like the following:

`FROM ubuntu:15.04`
`ENV AP /app`
`WORKDIR $AP/`
`ENV LD_LIBRARY_PATH $AP/`
`ADD dockerDependencies/*.* $AP/`
`ADD mybinary $AP/`

Note the setting of the `LD_LIBRARY_PATH` environment variable. This was part of my undoing... However, it seemed necessary, since the mybinary dependencies (which had been extracted via `ldd`) were being copied right next to it...

On my local VM, based on Ubuntu 15.04, running this presented no problem. Also, on a test VM it worked fine. The problems started when I was trying to automate the builds using a containerized GCC and collecting the output binaries. Suddenly, the containers built this way started failing with SIGSEGV.

I tried winnowing down the scripts to something that might build correctly and I thought I had it when I removed the WORKDIR line. Stupidly enough, I had also removed the `ENV LD_LIBRARY_PATH $AP` line and didn't notice and proceeded to blame Docker for the crashes. Shame on me...

To cut a long story short, the root of the issue was that I was copying dependency libraries from the GCC container and also getting some of the container's system libraries as a result. Since the container was built `FROM buildpack-deps:jessie` it was getting that build's core libraries such as: `ld-linux-x86-64.so.2` and `libc.so.6`. It turns out these were incompatible with the core libraries of the subsequent container used to run the application. So when I used the  `ENV LD_LIBRARY_PATH $AP/` directive, I was forcing a Ubuntu container to use Jessie core binaries. In hindsight, SIGSEGV was to be expected...

Sorry for the wrongful post, lesson learned.
